### ISSUE TITLE
pseudo code

### ISSUE THREAD

**Issue by jdakka** [radical-collaboration/MSKCC#5]
@pgrinaway could you provide or point to the pseudocode for your Celery application, either here or on /perses (wherever you're more comfortable posting)

**Comment by pgrinaway** [comment]
Ok, though I'm still having some trouble so this might change. I was hoping I'd solve it tonight, but no such luck.

Main loop:
```
given thermodynamic_state, sampler_state
for iteration in iterations:
    task_chain = eq_simulation(thermodynamic_state, sampler_state) | neq_simulation, eq_output
    result = task_chain()
    eq_result = result[1] #eq_output
    thread_record_queue.put(result) #another thread waits on the results of neq_simulation
    sampler_state = eq_result.sampler_state
```

Listener thread:
```
given (potentially empty) float arrays work_values, endpoint_perturbations
while true:
   task= thread_record_queue.get()
   eq_result = task[1]
   neq_result = task[0] #this blocks
   
    reduced_potential_na = reduced_potential(eq_result) #this is a celery task
    endpoint_perturbations.append(eq_result.reduced_potential - reduced_potential_na)

    work_values.append(neq_result.work)
```

celery workers:
```
def eq_simulation(...):
    do equilibrium simulation (GPU)
    write result to file
    return final state (not entire trajectory)

def neq_simulation(...):
    do nonequilibrium simulation (GPU) (longest step)
    write result to file
    return final work value (float)

def reduced_potential(...):
    point energy evaluation (quick, could be CPU or GPU)
    return point energy

def eq_output(...):
    dummy task to pipe the output of the equilibrium task that precedes it.
```

As you can see there are a few limitations of celery that I'm trying to work around. It doesn't seem to be well-supported to have complex nested groups, so I've had to merge the file writing into the compute tasks. I also need to use a dummy task because I don't think I can get the result of the first task in a chain otherwise. 

In general, it doesn't seem to be a big deal to need to return to the client for each equilibrium iteration, since the nonequilibrium ones are much longer.

**Comment by jdakka** [comment]
Can you elaborate on `eq_simulation(thermodynamic_state, sampler_state) | neq_simulation, eq_output`

From our previous discussions we laid out that eq and noneq would be the initial lambda states per phase. Can you be more specific on ```def eq_output(...):
    dummy task to pipe the output of the equilibrium task that precedes it.```

Also can you elaborate on the limitations: can you give a specific example of how you're using complex nested groups wrt eq/noneq states

**Comment by pgrinaway** [comment]
Ok, trying to go to a higher-level representation of the algorithm

For a defined set of iterations, in each of two `lambda` values (the algorithm is otherwise identical)
```
given hybrid_system, hybrid_state, endpoint_system
switch_work_values = []
endpoint_work_values = []

for each iteration:
    hybrid_state, end_potential = simulate_equilibrium(hybrid_system) #GPU
    switch_work_value = simulate_nonequilibrium(hybrid_state) #GPU
    nonhybrid_end_potential = calculate_end_potential(hybrid_state, hybrid_system, endpoint_system) #could be GPU or CPU
    
    switch_work_values.append(switch_work_value)
    endpoint_work_values.append(end_potential - nonhybrid_end_potential)
```

After running iterations:

```
#BAR analysis, ddeltaG is going to define the extent to which the calculation is converged
#this takes the results from both lambda simulations.
deltaG, ddeltaG = calculate_free_energies(switch_work_values, endpoint_work_values)
```


This might be missing specific details, so if you notice something, let me know and I'll try to add it in.

**Comment by pgrinaway** [comment]
Oh, and I just realized you asked about the pain points. Here's a summary thus far:

1) We need to be able to capture the output of the equilibrium and nonequilibrium simulations on the driver, both for output to the user and control. However, most frameworks that allow us to chain tasks make it difficult to get the result of a task farther back in the chain. Dask might help; I'll be chatting with someone in our group who has used it Monday.

2) We need to write files from the outputs of the GPU tasks. Some of these files can be quite large, so it would be best to do that writing in a separate task. However, that returns to problem 1, where we'd need the result of a task farther back in the task chain.
